OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameChatLogic.cs

```diff
@@ -263,53 +267,31 @@ public void AddChatLineWrapper(TextNotification chatLine)
 					chatLine.TextColor);
 
 				chatScrollPanel.RemoveChild(chatScrollPanel.Children[chatScrollPanel.Children.Count - 1]);
-				chatOverlayDisplay?.RemoveMostRecentLine();
+				chatOverlayDisplay?.RemoveMostRecentNotification();
 			}
 			else
 				repetitions = 0;
 
 			lastLine = chatLine;
 
-			chatOverlayDisplay?.AddLine(chatLineToDisplay);
+			chatOverlayDisplay?.AddNotification(chatLineToDisplay);
 
 			// HACK: Force disable the chat notification sound for the in-menu chat dialog
 			// This works around our inability to disable the sounds for the in-game dialog when it is hidden
 			AddChatLine(chatLineToDisplay, chatOverlay == null);
 		}
 
-		void AddChatLine(TextNotification chatLine, bool suppressSound)
+		void AddChatLine(TextNotification notification, bool suppressSound)
 		{
-			var template = chatTemplate.Clone();
-			var nameLabel = template.Get<LabelWidget>("NAME");
-			var textLabel = template.Get<LabelWidget>("TEXT");
-
-			var name = "";
-			if (!string.IsNullOrEmpty(chatLine.Prefix))
-				name = chatLine.Prefix + ":";
-
-			var font = Game.Renderer.Fonts[nameLabel.Font];
-			var nameSize = font.Measure(chatLine.Prefix);
-
-			nameLabel.GetColor = () => chatLine.PrefixColor;
-			nameLabel.GetText = () => name;
-			nameLabel.Bounds.Width = nameSize.X;
-
-			textLabel.GetColor = () => chatLine.TextColor;
-			textLabel.Bounds.X += nameSize.X;
-			textLabel.Bounds.Width -= nameSize.X;
-
-			// Hack around our hacky wordwrap behavior: need to resize the widget to fit the text
-			var text = WidgetUtils.WrapText(chatLine.Text, textLabel.Bounds.Width, font);
-			textLabel.GetText = () => text;
-			var dh = font.Measure(text).Y - textLabel.Bounds.Height;
-			if (dh > 0)
+			var chatLine = Ui.LoadWidget(templates[notification.Pool.ToString()], null, new WidgetArgs
```

Comment:
`Ui.LoadWidget` uses a significant amount of reflection, so i expect it to be orders of magnitude slower than `chatTemplate.Clone()`. Can we find a way to avoid this?

Reply:
Is something like this better:
```csharp
// In the ctor
if (logicArgs.TryGetValue("Templates", out var templateIds))
	foreach (var item in templateIds.ToDictionary(kv => kv.Value))
		templates[item.Key] = Ui.LoadWidget(item.Value, null, new WidgetArgs());

// and then when adding the lines
var chatLine = templates[notification.Pool.ToString()].Clone();
new TextNotificationLogic(chatLine, notification, boxWidth, withTimestamp);
```

I tested and it works but vscode is flagging the "new instance of TextNotificationLogic which is never used".

Reply:
Or maybe handle this with a static method on the `TextNotificationLogic` class (or elsewhere)? Something like:
```csharp
// when adding the lines
var chatLine = templates[notification.Pool.ToString()].Clone();
TextNotificationLogic.ApplyTo(chatLine, notification, boxWidth, withTimestamp);
```

Reply:
Added a static method to `WidgetUtils`.